pkg/cmd/roachtest/tests/mixed_version_cdc.go

```diff
@@ -0,0 +1,318 @@
+		t.Status("installing and running workload")
+		tpcc := tpccWorkload{
+			sqlNodes:           cmvt.crdbNodes,
+			workloadNodes:      cmvt.workloadNodes,
+			tpccWarehouseCount: 10,
+			tolerateErrors:     true, // we're bringing nodes down while upgrading
+		}
+
+		tpcc.install(ctx, u.c)
+		time.Sleep(2 * time.Second)
+		cmvt.monitor.Go(func(ctx context.Context) error {
+			tpcc.run(ctx, u.c, d)
+			return nil
+		})
+	}
+}
+
+// waitForVerifier waits for the underlying CDC verifier to reach the
+// desired number of resolved timestamps.
+func (cmvt *cdcMixedVersionTester) waitForVerifier() versionStep {
+	return func(ctx context.Context, t test.Test, u *versionUpgradeTest) {
+		t.Status("waiting for verifier")
+		<-cmvt.verifierDone
+	}
+}
+
+// setupVerifier creates a CDC validator to validate that a changefeed
+// created on the `tpcc.warehouse` table is able to re-create the
+// table somewhere else. It also verifies CDC's ordering
+// guarantees. This step will not block, but will start the verifier
+// in a separate Go routine. Use `waitForVerifier` to wait for
+// the verifier to finish.
+func (cmvt *cdcMixedVersionTester) setupVerifier(node int) versionStep {
+	return func(ctx context.Context, t test.Test, u *versionUpgradeTest) {
+		t.Status(fmt.Sprintf("setting up changefeed verifier for table %s", target))
+
+		// we could just return the error here and let `Wait` return the
+		// error. However, calling t.Fatal directly lets us stop the test
+		// earlier
+		cmvt.monitor.Go(func(ctx context.Context) error {
+			consumer, err := cmvt.kafka.consumer(ctx, "warehouse")
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer consumer.Close()
+
+			db := u.conn(ctx, t, node)
+			if _, err := db.Exec(
+				"CREATE TABLE fprint (" +
+					"w_id INT NOT NULL PRIMARY KEY, " +
+					"w_name VARCHAR(10) NOT NULL, " +
+					"w_street_1 VARCHAR(20) NOT NULL, " +
+					"w_street_2 VARCHAR(20) NOT NULL, " +
+					"w_city VARCHAR(20) NOT NULL, " +
+					"w_state VARCHAR(2) NOT NULL, " +
+					"w_zip VARCHAR(9) NOT NULL, " +
+					"w_tax DECIMAL(4, 4) NOT NULL, " +
+					"w_ytd DECIMAL(12, 2) NOT NULL)",
+			); err != nil {
+				t.Fatal(err)
+			}
+
+			fprintV, err := cdctest.NewFingerprintValidator(db, `tpcc.warehouse`, `fprint`, consumer.partitions, 0)
+			if err != nil {
+				t.Fatal(err)
+			}
+			validators := cdctest.Validators{
+				cdctest.NewOrderValidator("tpcc.warehouse"),
+				fprintV,
+			}
+			cmvt.validator = cdctest.MakeCountValidator(validators)
+
+			for {
+				m := consumer.Next(ctx)
+				if m == nil {
+					t.Fatal(fmt.Errorf("unexpected end of changefeed"))
+					return nil
+				}
+
+				updated, resolved, err := cdctest.ParseJSONValueTimestamps(m.Value)
+				if err != nil {
+					t.Fatal(err)
+					return nil
+				}
+
+				partitionStr := strconv.Itoa(int(m.Partition))
+				// errors in the calls to the validator below could lead to
+				// transient failures if the cluster is upgrading, so we retry
+				// for 1 minute.
```

Comment:
how often do these errors happen? if upgrade is happening, does it mean that we'll be spending up to 1 minute per row?

Reply:
Not per row, because we're not processing rows concurrently. These transient errors can last more than a minute though; I had to increase the timeout here to 2 minutes.